### PR TITLE
add prometheus metrics endpoint

### PR DESCRIPTION
**Pull request by sven-petersen**
**What this PR does / why we need it**:

Adds a second http server with a `/metrics` endpoint that can be scraped by prometheus.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

Tests are still a WIP

**Release note**:

```feature developer
Added monitoring endpoint for a prometheus scraper
```

### CODE REVIEW

**Comment by gardener-robot-ci-3** [comment]
Thank you @sven-petersen for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: Dockerfile [hunk 1]
```diff
@@ -79,11 +79,14 @@ ENV NODE_ENV "production"
 ARG PORT=8080
 ENV PORT $PORT
 
+ARG METRICSPORT=3040
+ENV METRICSPORT $METRICSPORT
```
Comment (petersutter) [inline Dockerfile:83]: split words with _ `METRICS_PORT`

File: backend/lib/server.js [hunk 2]
```diff
@@ -21,19 +21,22 @@ function toMilliseconds (seconds) {
...
+  const healthCheckFunc = app.get('healthCheck')
   const logger = app.get('logger')
   const hooks = app.get('hooks')
 
   // create server
   const server = http.createServer(app)
 
   // create terminus
+  const healthChecks = {}
+  if (typeof healthCheckFunc === 'function') {
```
Comment (petersutter) [inline backend/lib/server.js:33]: why do we need this check?

File: charts/gardener-dashboard/charts/runtime/templates/deployment.yaml [hunk 3]
```diff
@@ -99,6 +99,9 @@ spec:
             - name: http
               containerPort: {{ .Values.global.containerPort }}
               protocol: TCP
+            - name: metrics
+              containerPort: {{ .Values.global.metricsContainerPort }}
+              protocol: TCP
```
Comment (petersutter) [inline charts/gardener-dashboard/charts/runtime/templates/deployment.yaml:104]: don't we also need the `prometheus.io/scrape`, `prometheus.io/port` annotation?

File: packages/monitor/lib/app.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  const logger = new Logger()
+
+  const app = express()
+
+  app.set('port', port)
+  app.set('periodSeconds', periodSeconds)
+  app.set('hooks', hooks)
+  app.set('logger', logger)
+
+  app.set('trust proxy', 1)
```
Comment (petersutter) [inline packages/monitor/lib/app.js:25]: I don't think we need this setting for the monitoring package

File: charts/gardener-dashboard/charts/runtime/templates/deployment.yaml (hunk 3 above)
Comment (sven-petersen) [reply charts/gardener-dashboard/charts/runtime/templates/deployment.yaml:104]: the scrape job will look for our `app.kubernetes.io/name` and `.../component` as well as a port named `metrics` instead.

File: backend/lib/server.js (hunk 2 above)
Comment (sven-petersen) [reply backend/lib/server.js:33]: The API/backend and metrics apps are both passed to this function for creating the server (this approach is open for debate). For the metrics endpoint we do not have/want a health check. So to make this `createServer`-logic reusable we only add the health check routes if the `app` has defined a health check function.

**Comment by sven-petersen** [comment]
@holgerkoser @petersutter let's discuss how we want to name the configurations for the pod name & namespace.

File: packages/monitor/package.json [hunk 5]
```diff
@@ -0,0 +1,77 @@
+{
```
Comment (holgerkoser) [inline packages/monitor/package.json:1]:
```sh
(⎈ |N/A:default)❯ yarn constraints
➤ YN0024: @gardener-dashboard/backend must depend on express via ^4.18.2, but uses ^4.18.1 instead (in dependencies)
➤ YN0024: @gardener-dashboard/backend must depend on helmet via ^6.0.1, but uses ^5.1.0 instead (in dependencies)
➤ YN0024: @gardener-dashboard/backend must depend on supertest via ^6.3.3, but uses ^6.2.3 instead (in devDependencies)
➤ YN0024: @gardener-dashboard/kube-client must depend on express via ^4.18.2, but uses ^4.18.1 instead (in devDependencies)
➤ YN0024: @gardener-dashboard/monitor must depend on express via ^4.18.1, but uses ^4.18.2 instead (in dependencies)
➤ YN0024: @gardener-dashboard/monitor must depend on helmet via ^5.1.0, but uses ^6.0.1 instead (in dependencies)
➤ YN0024: @gardener-dashboard/monitor must depend on supertest via ^6.2.3, but uses ^6.3.3 instead (in dependencies)
➤ YN0000: Failed with errors in 0s 660m
```

File: charts/gardener-dashboard/values.yaml [hunk 6]
```diff
@@ -50,6 +50,7 @@ global:
     # # the identifier of the gardener landscape (defaults to the name stored in kube-system/cluster-identity configmap)
     # clusterIdentity: my-landscape-dev
     containerPort: 8080
+    metricsContainerPort: 3040
```
Comment (holgerkoser) [inline charts/gardener-dashboard/values.yaml:53]:
```suggestion
    metricsContainerPort: 9050
```

File: packages/monitor/lib/monitors.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const { metrics } = require('./metrics')
+
+function createWriteHead (prevWriteHead, listener) {
```
Comment (holgerkoser) [inline packages/monitor/lib/monitors.js:11]: Use http://expressjs.com/en/resources/middleware/response-time.html instead

File: packages/monitor/lib/monitors.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    metrics.connectionsTotal.inc({ type: 'ws' }, 1)
+  })
+}
+
+function monitorHttpServer (server) {
+  server.on('connect', () => {
+    metrics.connectionsCount.inc({ type: 'http' }, 1)
+    metrics.connectionsTotal.inc({ type: 'http' }, 1)
+  })
+  server.on('close', () => {
```
Comment (holgerkoser) [inline packages/monitor/lib/monitors.js:57]: This is emitted when the server closes. 
https://nodejs.org/api/http.html#event-close_1

File: packages/monitor/lib/monitors.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    socket.once('disconnect', () => {
+      metrics.connectionsCount.dec({ type: 'ws' }, 1)
+    })
+    metrics.connectionsCount.inc({ type: 'ws' }, 1)
+    metrics.connectionsTotal.inc({ type: 'ws' }, 1)
+  })
+}
+
+function monitorHttpServer (server) {
+  server.on('connect', () => {
```
Comment (holgerkoser) [inline packages/monitor/lib/monitors.js:53]: This event is emitted on HTTP CONNECT requests. This is not what we want. 
https://nodejs.org/api/http.html#event-connect_1

File: packages/monitor/lib/monitors.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+function monitorSocketIO (io) {
+  io.on('connection', (socket) => {
+    socket.once('disconnect', () => {
+      metrics.connectionsCount.dec({ type: 'ws' }, 1)
+    })
+    metrics.connectionsCount.inc({ type: 'ws' }, 1)
+    metrics.connectionsTotal.inc({ type: 'ws' }, 1)
+  })
```
Comment (holgerkoser) [inline packages/monitor/lib/monitors.js:49]:
```suggestion
  const labels = { type: 'ws' }
  io.on('connection', socket => {
    socket.once('disconnect', () => {
      metrics.connectionsCount.labels(labels).dec()
    })
    metrics.connectionsCount.labels(labels).inc()
    metrics.connectionsTotal.labels(labels).inc()
  })
```

File: packages/monitor/lib/monitors.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  server.on('connect', () => {
+    metrics.connectionsCount.inc({ type: 'http' }, 1)
+    metrics.connectionsTotal.inc({ type: 'http' }, 1)
+  })
+  server.on('close', () => {
+    metrics.connectionsCount.dec({ type: 'http' }, 1)
+  })
+}
+
+module.exports = { monitorResponseTimes, monitorSocketIO, monitorHttpServer }
```
Comment (holgerkoser) [inline packages/monitor/lib/monitors.js:62]:
```suggestion
module.exports = { 
  monitorResponseTimes, 
  monitorSocketIO, 
  monitorHttpServer
}
```

File: packages/monitor/lib/metrics.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    buckets: [
+      ...prometheus.linearBuckets(0, 0.1, 5), // 0-0.5s with 100ms buckets
+      ...prometheus.linearBuckets(0.5, 0.5, 5), // 0.5-3s with 500ms buckets
+      ...prometheus.linearBuckets(3, 1, 12) // 3-15s with 1000ms buckets
+    ],
+    registers: [register]
+  })
+}
+
+function start () {
```
Comment (holgerkoser) [inline packages/monitor/lib/metrics.js:39]: Do we need this start stop here? Why not do this in hooks?

File: packages/monitor/lib/hooks.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const metrics = require('./metrics')
+
+class LifecycleHooks {
+  async beforeListen (server) {
+    metrics.start()
```
Comment (holgerkoser) [inline packages/monitor/lib/hooks.js:13]:
```suggestion
    prometheus.collectDefaultMetrics()
```

File: packages/monitor/lib/hooks.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  }
+
+  async cleanup () {
+    metrics.stop()
```
Comment (holgerkoser) [inline packages/monitor/lib/hooks.js:17]:
```suggestion
    prometheus.register.clear()
```

File: packages/monitor/lib/hooks.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
+//
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const metrics = require('./metrics')
```
Comment (holgerkoser) [inline packages/monitor/lib/hooks.js:9]:
```suggestion
const prometheus = require('prom-client')
```

File: packages/monitor/lib/route.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+//
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const express = require('express')
+const { getMetrics, contentType } = require('./metrics')
```
Comment (holgerkoser) [inline packages/monitor/lib/route.js:10]:
```suggestion
const { getMetrics } = require('./metrics')
const { contentType } = require('prom-client').register
```

File: packages/monitor/lib/route.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+const router = express.Router()
+
+router.route('/')
+  .get(async (req, res, next) => {
+    try {
+      res.set('Cache-Control', 'no-store, no-cache, must-revalidate, proxy-revalidate')
+      res.setHeader('content-type', contentType)
+      res.status(200).end(await getMetrics())
```
Comment (holgerkoser) [inline packages/monitor/lib/route.js:19]:
```suggestion
      res.set({
        'cache-control', 'no-store, no-cache, must-revalidate, proxy-revalidate',
        'content-type', contentType
      })
      res.send(await getMetrics())
```

File: packages/monitor/lib/app.js [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const express = require('express')
+const helmet = require('helmet')
+const route = require('./route')
+const hooks = require('./hooks')
+const { Logger } = require('@gardener-dashboard/logger')
```
Comment (holgerkoser) [inline packages/monitor/lib/app.js:13]: Why not use the global logger or inject the logger as option?

File: packages/monitor/lib/app.js [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+  app.use(helmet.noSniff())
+  app.use(helmet.hsts())
+
+  app.use('/metrics', route)
+  app.use('*', (req, res) => res.sendStatus(404))
+  app.use((err, req, res, next) => {
+    logger.error('Error in monitoring server: %s', err)
+    res.sendStatus(500)
+  })
```
Comment (holgerkoser) [inline packages/monitor/lib/app.js:35]:
```suggestion
app.use((req, res, next) => next(createError(404)))
app.use(function(err, req, res, next) {
  const { message, status = 500 } = err
  logger.error('Error in monitoring server: %s', message)
  res.status(status).json({ 
    status,
    message
  })
})
```

File: packages/monitor/lib/app.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+  app.set('periodSeconds', periodSeconds)
+  app.set('hooks', hooks)
+  app.set('logger', logger)
+
+  app.set('x-powered-by', false)
+
+  app.use(helmet.noSniff())
+  app.use(helmet.hsts())
+
+  app.use('/metrics', route)
```
Comment (holgerkoser) [inline packages/monitor/lib/app.js:30]:
```suggestion
  app.get('/metrics', (req, res, next) => {
    try {
          res.set({
        'cache-control', 'no-store, no-cache, must-revalidate, proxy-revalidate',
        'content-type', contentType
      })
      res.send(await getMetrics())
    } catch (err) {
      next(err)
    }
  })
```

File: packages/monitor/lib/metrics.js [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const prometheus = require('prom-client')
+
+const register = new prometheus.Registry()
```
Comment (holgerkoser) [inline packages/monitor/lib/metrics.js:11]: Why not use the default register?

File: packages/monitor/package.json [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+    "restoreMocks": true,
+    "verbose": true,
+    "coverageDirectory": "coverage",
+    "testEnvironment": "node",
+    "coverageThreshold": {
+      "global": {
+        "branches": 69,
+        "functions": 100,
+        "lines": 96,
+        "statements": 96
```
Comment (holgerkoser) [inline packages/monitor/package.json:66]:
```suggestion
        "branches": 100,
        "functions": 100,
        "lines": 100,
        "statements": 100
```

**Comment by petersutter** [comment]
from the latest build failure log:
```
#12 [builder  7/16] RUN yarn install --immutable --immutable-cache
#12 sha256:140ab1b4b91a3f3caa938fd7b6ef4e51343c7a4c760802a288bb831c778a0670
#12 1.596 ➤ YN0000: ┌ Resolution step
#12 1.781 ➤ YN0001: │ Error: @gardener-dashboard/monitor@workspace:*: Workspace not found (@gardener-dashboard/monitor@workspace:*)
#12 1.781     at ze.getWorkspaceByDescriptor (/app/.yarn/releases/yarn-3.2.1.cjs:442:3273)
#12 1.781     at Ed.getCandidates (/app/.yarn/releases/yarn-3.2.1.cjs:395:29907)
#12 1.781     at Bd.getCandidates (/app/.yarn/releases/yarn-3.2.1.cjs:396:1281)
#12 1.781     at Bd.getCandidates (/app/.yarn/releases/yarn-3.2.1.cjs:396:1281)
#12 1.781     at /app/.yarn/releases/yarn-3.2.1.cjs:442:7764
#12 1.781     at Rg (/app/.yarn/releases/yarn-3.2.1.cjs:395:11098)
#12 1.781     at le (/app/.yarn/releases/yarn-3.2.1.cjs:442:7744)
#12 1.781     at async Promise.allSettled (index 4)
#12 1.781     at async uo (/app/.yarn/releases/yarn-3.2.1.cjs:395:10390)
#12 1.781 ➤ YN0000: └ Completed
#12 1.782 ➤ YN0000: Failed with errors in 0s 188ms
#12 ERROR: executor failed running [/bin/sh -c yarn install --immutable --immutable-cache]: exit code: 1
```

**Comment by holgerkoser** [comment]
> from the latest build failure log:
> 
> ```
> #12 [builder  7/16] RUN yarn install --immutable --immutable-cache
> #12 sha256:140ab1b4b91a3f3caa938fd7b6ef4e51343c7a4c760802a288bb831c778a0670
> #12 1.596 ➤ YN0000: ┌ Resolution step
> #12 1.781 ➤ YN0001: │ Error: @gardener-dashboard/monitor@workspace:*: Workspace not found (@gardener-dashboard/monitor@workspace:*)
> #12 1.781     at ze.getWorkspaceByDescriptor (/app/.yarn/releases/yarn-3.2.1.cjs:442:3273)
> #12 1.781     at Ed.getCandidates (/app/.yarn/releases/yarn-3.2.1.cjs:395:29907)
> #12 1.781     at Bd.getCandidates (/app/.yarn/releases/yarn-3.2.1.cjs:396:1281)
> #12 1.781     at Bd.getCandidates (/app/.yarn/releases/yarn-3.2.1.cjs:396:1281)
> #12 1.781     at /app/.yarn/releases/yarn-3.2.1.cjs:442:7764
> #12 1.781     at Rg (/app/.yarn/releases/yarn-3.2.1.cjs:395:11098)
> #12 1.781     at le (/app/.yarn/releases/yarn-3.2.1.cjs:442:7744)
> #12 1.781     at async Promise.allSettled (index 4)
> #12 1.781     at async uo (/app/.yarn/releases/yarn-3.2.1.cjs:395:10390)
> #12 1.781 ➤ YN0000: └ Completed
> #12 1.782 ➤ YN0000: Failed with errors in 0s 188ms
> #12 ERROR: executor failed running [/bin/sh -c yarn install --immutable --immutable-cache]: exit code: 1
> ```

The new package was not added to `.dockerignore`.

File: charts/gardener-dashboard/charts/runtime/templates/dashboard/deployment.yaml [hunk 23]
```diff
@@ -167,6 +170,8 @@ spec:
           - name: KUBECONFIG
             value: /etc/gardener-dashboard/secrets/kubeconfig/kubeconfig
           {{- end }}
+          - name: METRICS_PORT
+            value: {{ .Values.global.metricsContainerPort }}
```
Comment (petersutter) [inline charts/gardener-dashboard/charts/runtime/templates/dashboard/deployment.yaml:174]:
```suggestion
            value: {{ .Values.global.dashboard.metricsContainerPort }}
```

File: backend/lib/routes/shoots.js [hunk 24]
```diff
@@ -69,6 +75,7 @@ router.route('/:name')
   })
 
 router.route('/:name/spec/kubernetes/enableStaticTokenKubeconfig')
+  .all(metricsRoute('shoots/:name/spec/kubernetes/enableStaticTokenKubeconfig'))
```
Comment (petersutter) [inline backend/lib/routes/shoots.js:78]:
```suggestion
  .all(metricsMiddleware)
```

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by grolu — APPROVED** [review]
/lgtm

**Pull request merged**